Added count param to phrase param and return plural form accordingly

ZhenjaHorbach opened the pull request:
### Details


This PR added a count parameter to the phrase parameter to return the plural form when necessary. Also In this PR we 
 updated the types to support PluralizeValues and refactored the code so that all translation functions now use a single object containing all the necessary parameters

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
https://github.com/Expensify/App/issues/38614
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Open App
2. Open few screens and verify that the translations are displayed correctly.

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

1. Open App
2. Open few screens and verify that the translations are displayed correctly.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Open App
2. Open few screens and verify that the translations are displayed correctly.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/689b6246-38ed-4817-bc4d-561014d5fe46

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/8913de4b-3e41-4948-9316-a3b35b17dee8

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/2566276d-c473-42c8-8ac6-4983a2641a07

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/bfbd100f-9cfe-40f7-9ad9-db1bdb6dd5d1

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/02009ece-be40-4b89-97e2-87452d3af9a7

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/c8b5f7d1-68ca-46d6-8a4c-e9c641df7878

</details>

ZhenjaHorbach committed:
update TranslationBase and make few changes

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
update TranslationBase

ZhenjaHorbach committed:
update translations PART-1

ZhenjaHorbach committed:
update translations PART-2

ZhenjaHorbach committed:
update translations PART-3

ZhenjaHorbach committed:
update translations PART-4

ZhenjaHorbach committed:
Merge branch 'main' into pluralizing_localization-v2

ZhenjaHorbach committed:
update TranslationBase type

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
fix conflicts

dubielzyk-expensify commented:
Assume this needs no design review? If so, let me know and I'll check it out :)

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
update new translations

ZhenjaHorbach committed:
update TranslationBaseValue

ZhenjaHorbach committed:
revert some changes

ZhenjaHorbach committed:
revert some changes

ZhenjaHorbach committed:
update type arguments

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
fix ts issues

ZhenjaHorbach committed:
fix ts issue with no string values

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
update types for translations

blazejkustra commented on src/languages/en.ts:
```diff
@@ -4645,4 +4703,6 @@ export default {
-} satisfies TranslationBase;
+};
+
+export default translations satisfies TranslationBase<typeof translations>;
```
Okay now I understand, you moved the export under here for a reason hahah

If you try this:
```ts
const en = {
    test: 'lol',
    common: {
        yes: 'Yes',
        x: ({x}: {x: string}) => `should work ${x}`,
        j: ({test}: {test: string}, {test2}: {test2: string}) => `should fail, two argument not allowed ${test} ${test2}`,
        i: (num: number) => `should fail, argument is not an object ${num} ${ts}`,
        l: (num: Set<number>) => `should fail, argument is not an object ${num} ${ts}`,
        k: (x) => `should fail, argument is not an object ${x}`,
    },
} satisfies TranslationBase2<typeof en>;
```

You get an error:
`'en' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.`

ZhenjaHorbach replied:
Yeah 
I tried different ways
But my current version is the only one that works 😅

ZhenjaHorbach replied:
Great idea
Thanks !
But for better understanding
I added many conditions
To make ts error text relevant to the issue itself

blazejkustra replied:
I tried to simplify it, would that work? Do you see any bugs?

```ts
/**
 * Retrieves the first argument of a function
 */
type FirstArgument<TFunction> = TFunction extends (arg: infer A, ...args: any[]) => any ? A : never;

/**
 * Translation value can be a string or a function that returns a string
 */
type TranslationLeafValue<TStringOrFunction> = TStringOrFunction extends string
    ? string
    : (
          arg: FirstArgument<TStringOrFunction> extends Record<string, unknown> ? FirstArgument<TStringOrFunction> : Record<string, unknown>,
          ...noOtherArguments: unknown[]
      ) => string | PluralForm;

/**
 * Translation object is a recursive object that can contain other objects or string/function values
 */
type TranslationDeepObject<TTranslations> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [Path in keyof TTranslations]: TTranslations[Path] extends string | ((...args: any[]) => any) ? TranslationLeafValue<TTranslations[Path]> : TranslationDeepObject<TTranslations[Path]>;
};

const en2 = {
    test: 'lol',
    common: {
        yes: 'Yes',
        x: ({x}: {x: string}) => `should work ${x}`,
        plural: () => ({
            zero: 'No expenses',
            one: 'One expense',
            other: (count: number) => `${count} expenses`,
        }),
        // j: ({test}: {test: string}, {test2}: {test2: string}) => `should fail, two argument not allowed ${test} ${test2}`,
        // i: (num: number) => `should fail, argument is not an object ${num}`,
        // l: (sett: Set<number>) => `should fail, argument is not an object ${sett.has(1)}`,
        k: (x) => `should fail, argument is not an object ${x}`,
    },
};

const x = en2 satisfies TranslationDeepObject<typeof en2>;
```

ZhenjaHorbach replied:
But it looks better anyway 😅

blazejkustra replied:
@ZhenjaHorbach Are you sure? It errors for me 😲 

![Screenshot 2024-09-11 at 19 34 09](https://github.com/user-attachments/assets/352282e1-87fa-468f-9b8d-c381151a0701)

ZhenjaHorbach replied:
Oh yeah 
Restart ts server 
And now I see the issue 😅
I continue the test !

blazejkustra replied:
I found one problem, it doesn't work when a default value is passed 🤔 
```ts
const en2 = {
    test: 'lol',
    common: {
           deleteExpense: ({count}?: {count: number} = {count: 1}) => `Delete ${count}`,
    },
};

const x = en2 satisfies TranslationDeepObject<typeof en2>;
```

ZhenjaHorbach replied:
It works interesting
If this is an optional parameter
Then everything is fine

<img width="790" alt="Снимок экрана 2024-09-11 в 19 42 27" src="https://github.com/user-attachments/assets/ea4c97f1-3f64-4e94-8d89-3570cf3c9932">

blazejkustra replied:
Yep, found a fix!
```ts
/**
 * Retrieves the first argument of a function
 */
type FirstArgument<TFunction> = TFunction extends (arg: infer A, ...args: any[]) => any ? A : never;

/**
 * Translation value can be a string or a function that returns a string
 */
type TranslationLeafValue<TStringOrFunction> = TStringOrFunction extends string
    ? string
    : (
          arg: FirstArgument<TStringOrFunction> extends Record<string, unknown> | undefined ? FirstArgument<TStringOrFunction> : Record<string, unknown>,
          ...noOtherArguments: unknown[]
      ) => string | PluralForm;

/**
 * Translation object is a recursive object that can contain other objects or string/function values
 */
type TranslationDeepObject<TTranslations> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [Path in keyof TTranslations]: TTranslations[Path] extends string | ((...args: any[]) => any) ? TranslationLeafValue<TTranslations[Path]> : TranslationDeepObject<TTranslations[Path]>;
};
```

ZhenjaHorbach replied:
```
const translations = {
    test: 'lol',
    common: {
        yes: 'Yes',
        k: ({value}: {value: string}) => `should fail, argument is not an object ${value}`,
        deleteExpense: (count?: string) => `Delete ${count}`,
    },
};

export default translations satisfies TranslationBase<typeof translations>;
```

If we pass an optional parameter
There will be no ts error here

ZhenjaHorbach replied:
Actually works great!

blazejkustra replied:
U sure? 😅 
![Screenshot 2024-09-11 at 19 52 49](https://github.com/user-attachments/assets/7f4f583c-853f-4d77-a8b9-9034c42d890c)

ZhenjaHorbach replied:
I  removed the comment 😅
VS code is not feeling well today

ZhenjaHorbach replied:
typecheck also passed successfully !

ZhenjaHorbach replied:
Merged 
@blazejkustra 
Thanks for helping !

@shubham1206agra 
You can test it too 😃
And if everything is good
Let's fill out the checklist !

ZhenjaHorbach committed:
refactor TranslationBase

blazejkustra commented:
> @blazejkustra And have you noticed its not giving any error on using Array as output? Example: `mmk: ['jjkkkjk', 'bb2'],`
> 
> And add a key in `en.ts`, and not in `es.ts`, and linter does not complain

Will look into it tomorrow morning, interesting. Have you found any other edge cases? @shubham1206agra

blazejkustra commented:
@ZhenjaHorbach Meanwhile could you have a look at this [commit](https://github.com/Expensify/App/pull/49016/commits/718b3e71a4212f5c9b95cebf7d5278b54c614f14#diff-b5550289349ce88302fe326429e7bb5b45f39d7e1ca9ac81c6320810fd7bf740)? I moved all params to a separate file params.ts, and renamed/added useful comments to existing type helpers. I think it would be nice to apply some of these changes here, what do you think?

ZhenjaHorbach commented:
> > @blazejkustra And have you noticed its not giving any error on using Array as output? Example: `mmk: ['jjkkkjk', 'bb2'],`
> > And add a key in `en.ts`, and not in `es.ts`, and linter does not complain
> 
> Will look into it tomorrow morning, interesting. Have you found any other edge cases? @shubham1206agra

Oh, sorry 
My fault 
Today I also noticed this issue 
Here is an example of my fix

https://github.com/Expensify/App/pull/48229/commits/554611ccf0914458a49e4eef0a342516b3fe958d#diff-d5ecc5286b0686d54ec1fb926d76a79a5d8ecaeaed8a2366c7c9ce37a1ffabecR298

ZhenjaHorbach commented:
> @ZhenjaHorbach Meanwhile could you have a look at this [commit](https://github.com/Expensify/App/pull/49016/commits/718b3e71a4212f5c9b95cebf7d5278b54c614f14#diff-b5550289349ce88302fe326429e7bb5b45f39d7e1ca9ac81c6320810fd7bf740)? I moved all params to a separate file params.ts, and renamed/added useful comments to existing type helpers. I think it would be nice to apply some of these changes here, what do you think?

I will check today or tomorrow morning !

ZhenjaHorbach committed:
refactor some code

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach commented:
> @ZhenjaHorbach Meanwhile could you have a look at this [commit](https://github.com/Expensify/App/pull/49016/commits/718b3e71a4212f5c9b95cebf7d5278b54c614f14#diff-b5550289349ce88302fe326429e7bb5b45f39d7e1ca9ac81c6320810fd7bf740)? I moved all params to a separate file params.ts, and renamed/added useful comments to existing type helpers. I think it would be nice to apply some of these changes here, what do you think?

About this 
Changes looks good !
But I see we have a couple of errors
If we have an object with params that has optional values
I think it's okay if we don't pass the object itself

At the moment we are required to pass an empty object
`title={translate('iou.deleteExpense', {})}`


<img width="762" alt="Снимок экрана 2024-09-12 в 14 32 40" src="https://github.com/user-attachments/assets/7d2445a2-4ce2-4fdd-a934-fac4b9ee0fd0">

ZhenjaHorbach committed:
refactor types for translations

blazejkustra commented:
@ZhenjaHorbach fixed it in the latest [commits on my PR](https://github.com/Expensify/App/pull/49016)

ZhenjaHorbach committed:
update types

ZhenjaHorbach committed:
fix issue with no object values

blazejkustra commented:
@ZhenjaHorbach [One more commit to fix the last TS error](https://github.com/Expensify/App/pull/49016/commits/e010840f53bc5b78f091105b5f948b94cd40b023) - I adjusted the logic inside of `getTranslatedPhrase` function

blazejkustra commented on src/languages/types.ts:
```diff
@@ -1,501 +1,121 @@
+
+type TranslationParameters<TPath extends TranslationPaths> = FlatTranslationsObject[TPath] extends (...args: infer A) => unknown ? A : never[];
+
+/**
+ * Check all translations that are functions to make sure they have a valid argument
+ *
+ * The argument must be an defined object or undefined
+ *
+ * Example:
+ * {
+ *   // This is valid because the argument is defined
+ *   content: ({content}: ReportContentArgs) => `This is the content: ${content}`,
+ *
+ *   // This is invalid because the argument is not defined
+ *   content: ({content}) => `This is the content: ${content}`,
+ *
+ *   // This is invalid because the argument is not an object
+ *   content: (content: string) => `This is the content: ${content}`,
+ * }
+ */
+type InvalidFunctionTranslations = {
+    [Path in keyof FlatTranslationsObject]: FlatTranslationsObject[Path] extends infer Value
+        ? Value extends (...args: any[]) => any
+            ? Parameters<Value>[0] extends Record<string, unknown> | undefined
+                ? never
+                : Path
+            : never
+        : never;
+}[keyof FlatTranslationsObject];
+
+type TranslationArgumentShouldBeObjectError =
+    "ERROR: The argument of a translation function must be an object. The keys highlighted in 'InvalidFunctionTranslations' type have invalid function parameter.";
+
+/** If this type errors, it means that the translation functions have invalid arguments,  */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+type AssertTranslations = AssertTypesEqual<never, InvalidFunctionTranslations, TranslationArgumentShouldBeObjectError>;
```
This part is optional, we can remove it if you think it's redundant.

ZhenjaHorbach committed:
fix ts issue in localize

blazejkustra commented on src/languages/types.ts:
```diff
@@ -1,501 +1,121 @@
+    [Path in TranslationPaths]: TranslationValue<DefaultTranslation, Path>;
+};
+
+type TranslationParameters<TPath extends TranslationPaths> = FlatTranslationsObject[TPath] extends (...args: infer A) => unknown ? A : never[];
```
Let's add a comment for this utility too 🙏

blazejkustra reviewed (changes requested):
We are really close to the final solution, great job!

ZhenjaHorbach commented:
> We are really close to the final solution, great job!

I feel it 😀

Only this issue remains
https://github.com/Expensify/App/pull/48229#discussion_r1755276048

And this
https://github.com/Expensify/App/pull/48229#issuecomment-2344374348

First part i fixed 
But about second I am not sure 
Not sure how best to fix this
Because our objects with translations may differ slightly
Since the Spanish translation may have additional values (For example allCountries and allStates)

ZhenjaHorbach committed:
fix comments

blazejkustra commented:
> > We are really close to the final solution, great job!
> 
> I feel it 😀
> 
> Only this issue remains [#48229 (comment)](https://github.com/Expensify/App/pull/48229#discussion_r1755276048)
> 
> And this [#48229 (comment)](https://github.com/Expensify/App/pull/48229#issuecomment-2344374348)
> 
> First part i fixed But about second I am not sure how best to fix this Because our objects with translations may differ slightly Since the Spanish translation may have additional values (For example allCountries and allStates)

Let me look for a fix 👀

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
fix conflicts

blazejkustra commented:
Oh, I forgot to look into it today, will do first thing in the morning, sorry!

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
update package-lock

ZhenjaHorbach committed:
update package-lock x2

blazejkustra commented:
> And add a key in en.ts, and not in es.ts, and linter does not complain

```diff
diff --git a/src/languages/es.ts b/src/languages/es.ts
index 3454fd47dd4..3c66c5d20b4 100644
--- a/src/languages/es.ts
+++ b/src/languages/es.ts
@@ -1,5 +1,6 @@
 import {Str} from 'expensify-common';
 import CONST from '@src/CONST';
+import type en from './en';
 import type {
     AccountOwnerParams,
     ActionsAreCurrentlyRestricted,
@@ -194,7 +195,7 @@ import type {
 import type {TranslationDeepObject} from './types';
 
 /* eslint-disable max-len */
-const translations = {
+export default {
     common: {
         cancel: 'Cancelar',
         dismiss: 'Descartar',
@@ -5414,6 +5415,4 @@ const translations = {
         time: 'Hora',
         none: 'Ninguno',
     },
-};
-
-export default translations satisfies TranslationDeepObject<typeof translations>;
+} satisfies TranslationDeepObject<typeof en>;
```

blazejkustra commented:
> @blazejkustra And have you noticed its not giving any error on using Array as output?
Example: mmk: ['jjkkkjk', 'bb2'],

Can't reproduce it @shubham1206agra 🤔 It's giving me an error:
![image](https://github.com/user-attachments/assets/bd1c1c61-97ee-438f-af5e-2751db5a3b94)

blazejkustra commented:
Cool! This means all issues should be now resolved, let's aim to merge this bad boi finalyl 😄

ZhenjaHorbach committed:
fix comments

blazejkustra commented:
We can ignore `Changed files ESLint check` on this one 👍

ZhenjaHorbach committed:
fix conflicts

blazejkustra reviewed (approved):
LGTM!

blazejkustra commented:
> Since I may need to enforce count key to pluralized translations.

Could you give an example for it?

blazejkustra commented:
Both of these keys are in fact invalid because `other` has to be a function:
```ts
type PluralForm = {
    zero?: string;
    one: string;
    two?: string;
    few?: (count: number) => string;
    many?: (count: number) => string;
    other: (count: number) => string;
};
```

blazejkustra commented:
These are valid keys:
```ts
    key1: ({additionalProp}: {additionalProp: string}) => ({
        one: `1 b ${additionalProp}`,
        other: (count: number) => `${count} bs`,
    }),
    key2: () => ({
        one: '1 b',
        other: (count: number) => `${count} bs`,
    }),
```

You don't need to pass count as parameter because count is passed in the nested function to comply with `Intl.PluralRules`

blazejkustra commented on src/libs/Localize/index.ts:
```diff
@@ -82,39 +79,65 @@ const translationCache = new Map<ValueOf<typeof CONST.LOCALES>, Map<TranslationP
+                        return pluralResult;
+                    }
+
+                    return pluralResult(phraseObject.count);
```
I don't understand why is this problematic? 🤔

blazejkustra replied:
I think it was a design choice when we come up with `PluralForm` type, however we might be able to move the count to the outer function, so this is supported:
```ts
  key2: ({count}) => ({
        one: '1 b',
        other: `${count} bs`,
    }),
```

blazejkustra replied:
(I'm just helping with typescript side of things as this is quite complicated haha, design choice are yours @ZhenjaHorbach @shubham1206agra)

blazejkustra replied:
> We do need to pass count as a parameter otherwise it will fail to derive the correct value.

That isn't true, check this example, you don't need to pass count parameter in param types:

```ts
    key2: () => ({
        one: '1 b',
        other: (count: number) => `${count} bs`,
    }),
```

blazejkustra replied:
Why?

blazejkustra replied:
@shubham1206agra I think you are missing on the changes introduced in types.ts file, check out new type for parameters:
```ts
/**
 * Determines the expected parameters for a specific translation function based on the provided translation path
 */
type TranslationParameters<TPath extends TranslationPaths> = FlatTranslationsObject[TPath] extends (...args: infer Args) => infer Return
    ? Return extends PluralForm
        ? Args[0] extends undefined
            ? [PluralParams]
            : [Args[0] & PluralParams]
        : Args
    : never[];
```

ZhenjaHorbach committed:
fix conflicts

ZhenjaHorbach committed:
fix conflicts

iwiznia commented:
@ZhenjaHorbach can you check the 2 failed checks please?

ZhenjaHorbach committed:
update branch

iwiznia commented on src/libs/Localize/index.ts:
```diff
@@ -82,39 +79,65 @@ const translationCache = new Map<ValueOf<typeof CONST.LOCALES>, Map<TranslationP
+            if (phraseObject && typeof phraseObject === 'object' && 'count' in phraseObject && typeof phraseObject.count === 'number') {
+                const pluralRule = new Intl.PluralRules(language).select(phraseObject.count);
+
+                const pluralResult = pluralRule in translateResult && translateResult[pluralRule];
```
Do we need the `pluralRule in translateResult` I think the code would work exactly the same without it, no?

iwiznia commented on src/libs/Localize/index.ts:
```diff
@@ -82,39 +79,65 @@ const translationCache = new Map<ValueOf<typeof CONST.LOCALES>, Map<TranslationP
+                return translateResult.other(phraseObject.count);
+            }
+
+            throw new Error(`Invalid plural form for '${path}'`);
```
Throw early (ie: invert [this if](https://github.com/Expensify/App/pull/48229/files#diff-fe6e9abcee003c3390e38c0210d32ff889767cfab2c007837ed64c02147162e1R121) condition and throw)

iwiznia replied:
What all previous conditions? There's only one of them.
We prefer early throw/return in general, as it makes the code easier to read and follow

iwiznia reviewed (changes requested):
So we are removing the possibility to call `translate('bla', 'param', 'other')`. That's fine, but my fear is that other unmerged PRs could be using that format and we would not catch that, right?

Can you update the links in https://github.com/Expensify/App/#internationalization and add a small explanation of the pluralization feature?

iwiznia commented:
Oh, can you also pick some existing translation, for example [these ones](https://github.com/Expensify/App/blob/b079e3df2ee6a09d0240180c17069ec46747b433/src/languages/en.ts#L3589-L3591) and change them to use the new pluralization methods?
Just to leave as an example in real code.
I think I will create a new issue later to remove all usages of `Str.pluralize` which should not be needed anymore

iwiznia commented on src/libs/Localize/index.ts:
```diff
@@ -82,39 +79,65 @@ const translationCache = new Map<ValueOf<typeof CONST.LOCALES>, Map<TranslationP
+            }
+
+            const phraseObject = parameters[0];
+            if (phraseObject && typeof phraseObject === 'object' && 'count' in phraseObject && typeof phraseObject.count === 'number') {
```
Can't we change the type of `phraseObject` to allow for an optional `count`?

iwiznia replied:
> But I'm not sure that is a good idea update TranslationParameters type for count

Why?

> Plus we use such a condition to write count key into phraseObject

Which we could remove if we added it as an optional param, no?

ZhenjaHorbach committed:
fix some comments

ZhenjaHorbach committed:
update readme

ZhenjaHorbach committed:
update readme x2

ZhenjaHorbach committed:
update localize comment

iwiznia commented:
> Since in this case only TS will break on main after merging PR with wrong format of translates

Not sure I follow. Are you saying that if you call `translate(key, 'one', 'two')` it would still work? 

> This will be done by me in https://github.com/Expensify/App/pull/45892

Ah nice!

ZhenjaHorbach committed:
fix comments

iwiznia commented:
> Yes

Ah nice, thanks for checking and explaining!

iwiznia commented on src/languages/en.ts:
```diff
@@ -3661,7 +3661,10 @@ const translations = {
-            deleteRates: ({count}: DistanceRateOperationsParams) => `Delete ${Str.pluralize('rate', 'rates', count)}`,
+            deleteRates: () => ({
+                one: 'Delete rate',
+                other: () => `Delete rates`,
```
But same as the others, you could just have a string here right?

iwiznia commented on README.md:
```diff
@@ -619,7 +619,30 @@ Some pointers:
+
+  Here’s an example of how to implement plural translations:
+
+  messages: () => ({
+      zero: 'No messages',
+      one: 'One message',
+      two: 'Two messages',
+      few: (count) => `${count} messages`,
+      many: (count) => `You have ${count} messages`,
+      other: (count) => `You have ${count} unread messages`,
+  })
+
```
> No, since it will be impossible to differentiate between messages and messages.one translation key in a flat object.

I don't get this

> This implementation looks like any other object with translations
> And not a specific translation with plural forms

Yeah, so?

Anyway, it would be nice if we supported the simple object notation too, but not going to block on that since this is a big PR, we've been waiting a long time for it and adding the function is not that bad.

ZhenjaHorbach committed:
refactor plural forms

iwiznia commented:
We are skipping ESLint checks because this PR is touching 76 unrelated files and fixing all the previously existing lint errors would make this PR riskier.

Thanks so much for this, great job!!!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.
